src/diff/index.js

```diff
@@ -120,10 +120,14 @@ export function diff(
 				}
 
 				if (
-					!c._force &&
-					c.shouldComponentUpdate != null &&
-					c.shouldComponentUpdate(newProps, c._nextState, componentContext) ===
-						false
+					(!c._force &&
+						c.shouldComponentUpdate != null &&
+							c.shouldComponentUpdate(
+								newProps,
+								c._nextState,
+								componentContext
+							) === false) ||
+					(newVNode._original === oldVNode._original && !c._processingException)
```

Comment:
Q: could we hoist this to L39 to apply it to both component and element diffing at once?

Reply:
I think that will add a bunch off code sinceit will have to look like:

```
	if (newVNode._original === oldVNode._original) {
		const c = newVNode._component;
		if (c && !c._pendingError) {
			c.props = newVNode.props;
			c.state = c._nextState;
			c._dirty = false;
			c._vnode = newVNode;
			newVNode._dom = oldVNode._dom;
			newVNode._children = oldVNode._children;
			if (c._renderCallbacks.length) {
				commitQueue.push(c);
			}

			for (tmp = 0; tmp < newVNode._children.length; tmp++) {
				if (newVNode._children[tmp]) {
					newVNode._children[tmp]._parent = newVNode;
				}
			}

			return newVNode._dom;
		} else if (!c && excessDomChildren == null) {
			newVNode._children = oldVNode._children;
			newVNode._lastDomChild = oldVNode._lastDomChild;
			newVNode._dom = oldVNode._dom;
			return newVNode._dom;
		}
	}
```